external-crates/move/move-compiler/src/hlir/detect_dead_code.rs

```diff
@@ -13,6 +13,48 @@ use move_ir_types::location::*;
 use move_symbol_pool::Symbol;
 use std::{collections::BTreeMap, iter::Peekable};
 
+//**************************************************************************************************
+// Description
+//**************************************************************************************************
+// This analysis considers the input for potentially dead code due to control flow. It tracks
+// control flow in a somewhat fine-grained way, and when it finds a position that diverges it
+// reports that as an error.
+//
+// For simplicity, it aims to satify the following requirements:
+//
+//     1. For each block, if we discover a divergent instruction either at the top level or
+//        embedded in a value position (e.g., the RHS of a let, or tail-value position), we report
+//        that user to the error as possible dead code, under the following guidelines:.
+//        a) If the divergent code is nested within a value, we report it as a value error.
+//        b) If the divergent code is in a statement position and the block has a trailing unit as
+//           its last expression, report it as a trailing semicolon error.
+//        c) If the divergent code is in any other statement position, report it as such.
+//        d) If both arms of an if diverge in the same way in value or tail position, report the
+//           entire if together.
```

Comment:
This is new from after our conversation, right? 
Do you have a motivating example for this additional rule?

Reply:
It's just there to make the error a bit nicer -- happy to remove it.

Reply:
I like making this error nicer, it will probably be a relatively common case